crates/biome_service/src/workspace/scanner.rs

```diff
@@ -98,20 +99,52 @@ fn scan_folder(folder: &Utf8Path, ctx: ScanContext) -> (Duration, Vec<BiomePath>
         scope.evaluate(ctx_ref, folder.to_path_buf());
     }));
 
-    let evaluated_paths = ctx.evaluated_paths();
+    let interner = ctx.interner();
+    let dome = Dome::from_intern(interner.as_intern_set());
+    let mut iter = dome.iter();
 
     let mut configs = Vec::new();
     let mut manifests = Vec::new();
-    let mut handleable_paths = Vec::with_capacity(evaluated_paths.len());
+    let mut handleable_paths = Vec::new();
     let mut ignore_paths = Vec::new();
-    for path in evaluated_paths {
-        if path.is_config() {
-            configs.push(path);
-        } else if path.is_manifest() {
-            manifests.push(path);
-        } else if path.is_ignore() {
-            ignore_paths.push(path);
-        } else {
+
+    while let Some(path) = iter.next_config() {
+        configs.push(path.clone());
+    }
+
+    while let Some(path) = iter.next_manifest() {
+        manifests.push(path.clone());
+    }
+
+    while let Some(path) = iter.next_ignore() {
+        ignore_paths.push(path.clone());
+    }
+
+    let result = ctx
+        .workspace
+        .update_project_ignore_files(ctx.project_key, &ignore_paths);
+
+    for path in iter {
+        // If the path in inside `node_modules`, then we want to store only the `.d.ts` files,
+        // otherwise we don't store it at all
+        if path.is_dependency() {
+            if path.is_type_declaration() {
```

Comment:
```suggestion
            if path.is_package_json() || path.is_type_declaration() {
```

I don't think we have `is_package_json()` yet, but `is_manifest()` also returns `true` for `tsconfig.json`, and I think we have no use storing `tsconfig.json` files for dependencies.